### DISCUSSION TITLE
The g4f api cannot specify a provider! How to solve this?

### DISCUSSION THREAD

**Discussion by Perfecttoor** [xtekky/gpt4free#2333] [Q&A] [answered]
The g4f api cannot specify a provider! How to solve this?

**Comment by Linden10** [comment, marked as answer]
After looking into the code, I found out how to control which provider the api uses!

You're going to have to manually open models.py located somewhere in your g4f folder, usually located in the following locations:

System-Wide:
C:\Python[Version]\Lib\site-packages\g4f\models.py

Current User:
C:\Users\[user]\AppData\Local\Programs\Python\Python[Version]\Lib\site-packages\g4f\models.py

Open it in notepad or whatever default text editor you use and then scroll towards "Text" section near the top, there you'll see
various models with a "best_provider" list, look at the list of providers and remove anything you don't want to use for the given
ChatGPT Model you're using.

For example, if I wanted to remove ChatGptEs when using g4f api with gpt-4o, I'll edit it out like this:
```
# gpt-4
gpt_4o = Model(
    name          = 'gpt-4o',
    base_provider = 'OpenAI',
    best_provider = IterListProvider([NexraChatGPT, Blackbox, AmigoChat, DarkAI, Editee, GizAI, Airforce, Liaobots, OpenaiChat])
```

You can customize the list if it supports the given model, otherwise just remove what you don't want the api to use!

**Reply by Perfecttoor** [reply]
Amazing, thank you so much!

**Comment by hlohaus** [comment]
@Perfecttoor Also you can define one or multiple providers in the "provider" parameter: 

```python
import requests
import json
from g4f.image import to_data_uri
from g4f.requests.raise_for_status import raise_for_status

url = "http://localhost:8080/v1/chat/completions"
body = {
    "model": "",
    "provider": "You OpenaiAccount",
    "messages": [
        {"role": "user", "content": "what are on this image?"}
    ],
    "images": [
        [to_data_uri(open("docs/images/cat.jpeg", "rb")), "cat.jpeg"]
    ]
}
response = requests.post(url, json=body, headers={"g4f-api-key": "secret"})
raise_for_status(response)
print(response.json())
```
